queasars/job_shop_scheduling/scip_solver.py

```diff
@@ -0,0 +1,136 @@
+        :type solution: Solution
+        :return: a JobShopSchedulingResult parsed from the solution
+        :rtype: JobShopSchedulingResult
+        """
+        job_schedules: dict[Job, tuple[ScheduledOperation, ...]] = {}
+        for job in self._jssp_instance.jobs:
+            scheduled_operations: list[ScheduledOperation] = []
+            for operation in job.operations:
+                start_time: int = int(solution[self._operation_start_variables[operation]])
+                scheduled_operations.append(ScheduledOperation(operation=operation, start=start_time))
+            job_schedules[job] = tuple(scheduled_operations)
+
+        return JobShopSchedulingResult(problem_instance=self._jssp_instance, schedule=job_schedules)
+
+    def _prepare_variables(self):
+        """
+        Creates the integer variables for the starting times of each operation and constrains them to be at least 0.
+        Also creates a integer variable to hold the makespan, which is constrained to be larger than the end times
+        of the last operations. Sets the optimization goal to minimize the makespan variable
+        """
+        # create auxiliary variable to hold the makespan as an optimization goal
+        self._optimization_var = self._model.addVar("optimization_var", vtype="INTEGER")
+        self._model.setObjective(self._optimization_var)
+        self._model.addCons(self._optimization_var >= 0)
+
+        for job in self._jssp_instance.jobs:
+            for operation in job.operations:
+                # Cache the operations per machine
+                # This is needed for the additional constraints
+                if operation.machine not in self._machine_operations:
+                    self._machine_operations[operation.machine] = []
+                self._machine_operations[operation.machine].append(operation)
+
+                # Create an integer variable for the start time of each operation
+                # Constrain it to be at least 0
+                var = self._model.addVar(operation.identifier, vtype="INTEGER")
+                self._operation_start_variables[operation] = var
+                self._model.addCons(var >= 0)
+
+    def _prepare_constraints(self):
+        """
+        Constrains all operations within each job to start in order and not overlap.
+        Constrains all operations for each machine to not overlap. To this end auxiliary binary variables
+        are created to enable ordering decisions.
+        """
+        for job in self._jssp_instance.jobs:
+            n_operations = len(job.operations)
+
+            # Ensure by constraint, that each operation in a job finishes, before the next one may start
+            for i in range(0, n_operations - 1):
+                self._model.addCons(
+                    self._operation_start_variables[job.operations[i]] + job.operations[i].processing_duration
+                    <= self._operation_start_variables[job.operations[i + 1]]
+                )
```

Comment:
Here and below: You can give constraints a name such that debug and log messages are a bit more helpful: I typically write something like the type/name of the constraint plus an identifiert (e.g. if the constraint is per machine, I would add the machine name into the name).

Reply:
I added constraint descriptions in the name field. I hope they are not too verbose.